rfd/0160-kubernetes-operator-resource-versioning.md

```diff
@@ -0,0 +1,210 @@
+- The atypical versioning might be confusing for users as it doesn't follow
+  usual Kubernetes patterns.
+
+### API evolution
+
+This design unblocks us today, but does not cover cases in which we introduce a
+new resource version with a completely different structure with fields with
+different types. It is unclear if this can or will happen as this would very
+likely cause other issues in Teleport due to the way we handle resource versioning.
+
+If this were to happen, we would need to revise the solution from this RFD and
+introduce a new resource, or implement conversion hooks. We would also need to
+change the `resources.teleport.dev/vX` version to the current Teleport version.
+
+For example:
+- We implement this RFD in Teleport 15: the API is `resources.teleport.dev/v15`
+- We don't do any breaking changes before v20: the API is still `resources.teleport.dev/v15` for Teleport version 16, 17, 18 and 19.
+- In Teleport v20, we introduce a breaking change in Role. We introduce a new API `resoucres.teleport.dev/v20`.
+
+### Alternatives
+
+When writing this RFD, two alternatives were considered:
+
+#### Conversion hooks
+
+The Kubernetes-friendly approach would be to make the operator aware of how the
+resource is stored in Kubernetes, and do the conversions for every Kubernetes CR
+API call via webhooks.
+
+This alternative has been rejected because of the following reasons:
+- We currently handle resource conversion only from the old version to the newer versions,
+  this would require implementing bidirectional conversion between each resource version
+  (or use a Hub & Spoke approach by converting every version to a common version). This also 
+  raises questions about how to represent non-supported fields during downgrades (e.g. a
+  user gets a Role v7 through a Role v5 API). This would increase the cost of adding new
+  resources and increase the chances of addings bugs in the conversions due to the increasing
+  conversion matrix.
+- We are currently working on hiding `CheckAndSetDefaults()` from the client
+  and consolidating defaults injection and resource conversion server-side
+  This approach is not compatible with the conversion hooks pattern as we'd
+  need to run `CheckAndSetDefaults` in the operator and duplicate the logic.
+- This requires relying on conversion webhooks:
+  - This adds new failure modes: the operator is not healthy, kubernetes cannot talk to the operator
+  - Not all Kubernetes distributions support kube control plane to workload communication by default
+    (this is blocked on private GKE clusters, and most security teams also block this on their EKS clusters).
+    This would slow adoption and worsen UX.
+  - This requires the webhook server to be trusted by Kubernetes, which implies creating a CA, signing certs,
+    and putting them in the CRD resource. This complexifies the deployment process a lot.
+- The cost of implementing this solution is way higher than the other alternatives.
```

Comment:
Agreed but we recently saw the problems caused by cutting some corners with auto upgrades and that's why I am ultra afraid of cutting corners here

Reply:
That's not an issue if the layout is compatible which is not always the case.

https://github.com/gravitational/teleport/blob/dad8c4475cef5daea3075ba81c94291d280f4207/lib/services/provisioning.go#L79-L111

Here, you can see that we still support both V1 and V2 versions because the object doesn't expire and their manifest is incompatible.
Once you need to add a breaking change to the SAML connector (V3), you can't because the CRD is tied to v2 layout and you can't change it without breaking operator unless you keep the v2 schema forever.

This is not a huge problem in Teleport because 80% of the objects expire after some TTL but for kube_cluster, app, db, saml, oidc, users... we need to consider that otherwise we can't add layout changes and Teleport YAML and CRD will mismatch,

For those assets, usually we support marshaling and unmarshaling both types during 1 major version period and after 2 versions, we convert the older type to the new type in the backend. This is not possible in Kube because everything is managed externally, so we have several problems to consider